flink/v1.17/flink/src/main/java/org/apache/iceberg/flink/source/reader/IcebergTimestampWatermarkExtractor.java

```diff
@@ -0,0 +1,65 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.source.reader;
+
+import java.io.Serializable;
+import java.util.Comparator;
+import org.apache.flink.table.data.RowData;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.flink.source.split.IcebergSourceSplit;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+/**
+ * {@link IcebergWatermarkExtractor} implementation which uses an Iceberg timestamp column
+ * statistics to get the watermarks for the {@link IcebergSourceSplit}. This watermark is emitted by
+ * the {@link WatermarkExtractorRecordEmitter} along with the actual records.
+ */
+public class IcebergTimestampWatermarkExtractor
```

Comment:
I would keep either `Timestamp` or `Watermark`, not both. e.g. `IcebergSourceSplitWatermarkAssigner`

Reply:
This is the implementation of the `IcebergWatermarkExtractor` interface, and we use a `Timestamp` column for fetching the watermark.

I think it is more descriptive than `IcebergSourceSplitWatermarkAssigner`, as there could be `long` based extractors, or `int` based extractors....

Should this be `TimestampBasedWatermarkExtractor` instead?

Reply:
> I think it is more descriptive than IcebergSourceSplitWatermarkAssigner, as there could be long based extractors, or int based extractors....

what does this mean? this implementation class is about how to extract watermark (long) from a split. As mentioned in another comment, it should be `SplitWatermarkExtractor`.

> Should this be TimestampBasedWatermarkExtractor instead?

all watermarks are timestamp based. so this name probably doesn't make much sense.

Reply:
What I meant here is, that the user could decide, to store the timestamps in the table in a long column, or use an arbitrary method to calculate the watermark. Is there a restriction in Flink that the watermark should be timestamp based? My impression is that using a timestamp is a good practice, but any continuously increasing long number would be good too.

Reply:
For now implemented a Timestamp based watermark solution